src/Workspaces/Core/Portable/Execution/Serialization/Checksum.cs

```diff
@@ -73,8 +75,8 @@ public override int GetHashCode()
                     hash = Hash.Combine((int)_checkSum[i], hash);
                 }
 
-                _lazyHash = hash;
-            }
+                _lazyHash = new StrongBox<int>(hash);
```

Comment:
StrongBox won't fix the race condition :)

The race is that reads and writes can be reordered across threads in the absence of memory barriers.  As such, from the perspective of one thread _lazyHash may point to the newly allocated StrongBox, but the strongbox may not contain the hash you passed its constructor.  Again, writes can be reordered.  So the write of 'hash' to the field in StrongBox may not be seen, even those you see the write to _lazyHash.